File: omnivault-frontend/src/services/folderService.js
import axiosInstance from "./axiosInstance";
import { apiCache } from "../utils/apiCache";

const folderService = {
  getRootFolders: async () => {
    const response = await axiosInstance.get("/folders/root");
    return response.data;
  },

  getFolder: async (folderId) => {
    const cacheKey = `folder_${folderId}`;

    // Try to get from cache first
    const cachedFolder = apiCache.get(cacheKey);
    if (cachedFolder) {
      return cachedFolder;
    }

    // Not in cache, fetch from API
    const response = await axiosInstance.get(`/folders/${folderId}`);

    // Store in cache before returning
    apiCache.set(cacheKey, response.data);
    return response.data;
  },

  getSubfolders: async (folderId) => {
    const response = await axiosInstance.get(`/folders/${folderId}/subfolders`);
    return response.data;
  },

  createFolder: async (folderData) => {
    const response = await axiosInstance.post("/folders", folderData);
    return response.data;
  },

  updateFolder: async (folderId, folderData) => {
    const response = await axiosInstance.put(
      `/folders/${folderId}`,
      folderData
    );
    return response.data;
  },

  deleteFolder: async (folderId) => {
    await axiosInstance.delete(`/folders/${folderId}`);
  },

  searchFolders: async (query) => {
    const response = await axiosInstance.get(`/folders/search?query=${query}`);
    return response.data;
  },
};

export default folderService;

File: omnivault-backend/src/main/java/com/omnivault/service/FolderService.java
package com.omnivault.service;

import com.omnivault.domain.dto.request.FolderCreateRequest;
import com.omnivault.domain.dto.response.FolderDTO;
import com.omnivault.domain.model.Folder;

import java.util.List;
import java.util.UUID;

public interface FolderService {

    /**
     * Get all root folders for the current user
     *
     * @return List of root folders
     */
    List<FolderDTO> getRootFolders();

    /**
     * Get all subfolders of a folder
     *
     * @param folderId The parent folder ID
     * @return List of subfolders
     */
    List<FolderDTO> getSubfolders(UUID folderId);

    /**
     * Get folder by ID
     *
     * @param folderId The folder ID
     * @return The folder DTO
     */
    FolderDTO getFolder(UUID folderId);

    /**
     * Get folder entity by ID
     *
     * @param folderId The folder ID
     * @return The folder entity
     */
    Folder getFolderEntity(UUID folderId);

    /**
     * Create a new folder
     *
     * @param request The folder creation request
     * @return The created folder DTO
     */
    FolderDTO createFolder(FolderCreateRequest request);

    /**
     * Update a folder
     *
     * @param folderId The folder ID
     * @param request The folder update request
     * @return The updated folder DTO
     */
    FolderDTO updateFolder(UUID folderId, FolderCreateRequest request);

    /**
     * Delete a folder
     *
     * @param folderId The folder ID
     */
    void deleteFolder(UUID folderId);

    /**
     * Search folders by name or description
     *
     * @param searchTerm The search term
     * @return List of matching folders
     */
    List<FolderDTO> searchFolders(String searchTerm);
}